File: 33-search-in-rotated-sorted-array/search-in-rotated-sorted-array.cpp
class Solution {
public:
    int findMinElem(vector<int>& nums)
    {
        int l=0,h=nums.size()-1;
        while(l<h)
        {
            if(nums[l]<nums[h])
            return l;
            
            int mid=l+(h-l)/2;

            if(nums[l]<=nums[mid])
            l=mid+1;
            else
            h=mid;
        }
        return l;
    }
    int advancedBinSearch(vector<int>& nums,int l,int h,int x)
    {
        //int l=0,h=nums.size();
        while(l<h)
        {
            int mid=l+(h-l)/2;

            if(nums[mid]<x)
            l=mid+1;
            else
            h=mid;
        }
        return l;
    }
    int search(vector<int>& nums, int target) 
    {
        int l;
        int min_elem_idx=findMinElem(nums);
        
        if(target<=nums[nums.size()-1])
        l=advancedBinSearch(nums,min_elem_idx,nums.size(),target);
        //cout<<"Hi";
        else
        l=advancedBinSearch(nums,0,min_elem_idx-1,target);
        
        if(l==nums.size() || nums[l]!=target)
        return -1;

        return l;
    }
};

File: 33-search-in-rotated-sorted-array/search-in-rotated-sorted-array.java
class Solution {
    public int minElemPos(int nums[])
    {
        int l=0,h=nums.length-1;
        while(l<h)
        {
            int mid=l+(h-l)/2;
            if(nums[mid]>nums[h])
            l=mid+1;
            else
            h=mid;
        }
        return l;
    }
    public int advBinarySearch(int nums[],int low,int high,int target)
    {
        while(low<high)
        {
            int mid=low+(high-low)/2;

            if(nums[mid]<target)
            low=mid+1;
            else
            high=mid;
        }
        return low;
    }
    public int search(int[] nums, int target) {

        int pos_of_min_elem=minElemPos(nums);
        int pos;
        if(target<=nums[nums.length-1])
            pos=advBinarySearch(nums,pos_of_min_elem,nums.length,target);
        else
            pos=advBinarySearch(nums,0,pos_of_min_elem-1,target);
        
        if(pos==nums.length || nums[pos]!=target)
        return -1;
        else
        return pos;

    }
}